src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs

```diff
@@ -71,19 +89,194 @@ protected override void EndProcessing()
             {
                 return;
             }
-            orderByProperty.OrderMatrix.Sort(orderByProperty.Comparer);
 
-            if (_unique)
+            // If -Unique was used, if -Top & -Bottom were not used, or of -Top or -Bottom would return all objects, sort
+            // all objects, remove duplicates if necessary, and identify the list of items to return
+            if (_unique || (Top == 0 && Bottom == 0) || Top >= orderByProperty.OrderMatrix.Count || Bottom >= orderByProperty.OrderMatrix.Count)
             {
-                RemoveDuplicates(orderByProperty);
+                orderByProperty.OrderMatrix.Sort(orderByProperty.Comparer);
+
+                if (_unique)
+                {
+                    RemoveDuplicates(orderByProperty);
+                }
+
+                var listToOutput = orderByProperty.OrderMatrix;
+
+                if (Top > 0 && Top < listToOutput.Count)
+                {
+                    // There may be a faster algorithm to sort while eliminating duplicates and retaining only
+                    // the top N records; for now though, this gets the job done
+                    listToOutput = listToOutput.GetRange(0, Top);
+                }
+                else if (Bottom > 0 && Bottom < listToOutput.Count)
+                {
+                    int lastItemIndex = listToOutput.Count - 1;
+                    listToOutput = listToOutput.GetRange(lastItemIndex - Bottom, lastItemIndex);
+                }
+
+                // write to output stream
+                foreach (var x in listToOutput)
+                {
+                    WriteObject(x.inputObject);
+                }
+            }
+            // Otherwise, use an indexed min-/max-heap to sort all objects
+            else
+            {
+                // The heap entry comparer will sort the items in the heap
+                var comparer = new IndexedMinMaxHeapComparer(orderByProperty.Comparer);
+
+                // Identify how many items will be in the heap
+                int indexedMinMaxHeapSize = Top > 0 ? Top : Bottom;
+
+                // A small list will store the data for the heap 
+                var indexedMinMaxHeap = new List<IndexedMinMaxHeapEntry>(indexedMinMaxHeapSize + 1);
```

Comment:
@iSazonov: The algorithm I wrote is necessary and results in much faster performance than simply calling List.Sort and truncating the results. It is intentionally different because it uses a heap sort so that it gets the top/bottom N results faster than would be possible with List.Sort.

@lzybkr: The new list I create contains objects that include the original OrderByPropertyEntry object as well as its index. I track the index so that objects with the same sort order in the heap can be output in the order in which they were received from the unsorted list. This includes both sortable data items and unsortable data items. I'm looking at whether or not it's possible to preserve order in the heap without tracking the index; however, I don't believe it is (that was why I added the index and new *Entry class in the first place). Otherwise I would agree that I could update the logic to sort the list in place. I suppose I could track the original order in the OrderByPropertyEntry class (which could probably be struct as well, for the same reason you suggested it for my class) by adding an integer property for the index; however, that means changing classes/methods that I was intentionally leaving untouched. Thoughts?